### PR TITLE
Bump to mono:2019-06

### PR DESCRIPTION
**Pull request by directhex**
Bump to mono/2019-06 branch and set mono required versions.

[cecil] Fix the cecil build by reverting "Add a solution file for remap-assembly-ref (#3248)"

Using the solution instead of csproj broke the cecil build. I didn't
traced all the details, but we ended up with compiling
`Mono.Cecil.dll` without any references.

Use 7z when downloading Mono Archives 

[xaprepare] Drop llvm32 cross compilers

Mono has just dropped the llvm32 cross compilers because it is now possible to
cross compile to 32-bit using the 64-bit compiler.

Remove requirement for 32-bit llvm runtime in preparation stage.

[tests] Updated Xamarin.Android.Net tests

Update Properties_Invalid test to reflect current implementation

This was testing part of our old HttpClientHandler implementation
introduced here:
https://github.com/mono/mono/commit/f97e03f284f239e14ab308800679c1b2f0b04f64

The current corefx implementation doesn't throw when `UseProxy` is
false. The documentation doesn't specify such behavior, so I assume it
was specific for our old implementaion.

Context:
https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclienthandler.proxy?view=netframework-4.8
https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclienthandler.useproxy?view=netframework-4.8
https://github.com/mono/corefx/blob/9354dfd8d3cd4af62592bee9d8fe434e5694788a/src/System.Net.Http/src/System/Net/Http/HttpClientHandler.Unix.cs#L242
https://github.com/mono/corefx/blob/9354dfd8d3cd4af62592bee9d8fe434e5694788a/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocketsHttpHandler.cs#L75
https://github.com/mono/corefx/blob/9354dfd8d3cd4af62592bee9d8fe434e5694788a/src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.cs#L234

Remove broken HttpClient test

Context: https://github.com/xamarin/xamarin-android/pull/3155#issuecomment-506856300

[linker] Disable linker's UnreachableBodies optimization

Disable it for now til we find out a fix.

Context: https://github.com/xamarin/xamarin-android/issues/3202

[jnimarshalmethod-gen] Introduce MonoAndroidLibDirectory property

As we have switched to dynamically linker mono recently, we need to have
libmonosgen in dyld path. So we set `DYLD_LIBRARY_PATH` for `mono`
when running `jnimarshalmethod-gen`

[tests] Exclude network tests which hang.

Context: https://github.com/mono/mono/issues/14938
Context: https://github.com/xamarin/xamarin-macios/pull/6228

[xamarin-android-api-compatibility] API changes

System.IO.Enumeration.FileSystemName and
System.Security.Cryptography.CryptographicOperations classes were
removed from System.xml as they were duplicates of mscorlib ones. They
were removed as part of NS2.1 work in mono.

They were added by mistake and removed later. Because they were
already present in 16.1 and 16.2 respectively, mono team added type
forwarders to `System.dll` to avoid/lessen API breakage.

System.dll now contains:

	.class extern forwarder System.IO.Enumeration.FileSystemName
	{
	  .assembly extern mscorlib
	}
	.class extern forwarder System.Security.Cryptography.CryptographicOperations
	{
	  .assembly extern mscorlib
	}

Context:
https://github.com/mono/mono/commit/ba719acaa24e08e90b19df495a6c5a5cbe469803
https://github.com/xamarin/xamarin-android/pull/3155#issuecomment-510142890

3 extension methods extracted from
`System.Threading.Tasks.TaskExtensions` to a new
`System.Threading.Tasks.TaskAsyncEnumerableExtensions` class.

They stay in the same namespace and being extension methods, they
should not break anything.

	public static System.Runtime.CompilerServices.ConfiguredCancelableAsyncEnumerable<T> ConfigureAwait<T> (this System.Collections.Generic.IAsyncEnumerable<T>, bool);
	public static System.Runtime.CompilerServices.ConfiguredAsyncDisposable ConfigureAwait (this System.IAsyncDisposable, bool);
	public static System.Runtime.CompilerServices.ConfiguredCancelableAsyncEnumerable<T> WithCancellation<T> (this System.Collections.Generic.IAsyncEnumerable<T>, System.Threading.CancellationToken);

Context:
https://github.com/dotnet/standard/commit/ebeec50871621ac116b1257c791fc162a0ac5ea0#diff-93ef3714cd10edf17d6d8045cf3406b8
https://github.com/xamarin/xamarin-android/pull/3155#issuecomment-502672705

### CODE REVIEW

**Comment by jonpryor** [comment]
This [currently fails](https://jenkins.mono-project.com/job/xamarin-android-pr-pipeline-release/1107/) while trying to build `Mono.Cecil.dll`:

```
[2019-05-31T20:39:14.651Z] .../csc.exe /noconfig ... /out:obj/net_4_0_Debug/netstandard2.0/Mono.Cecil.dll ...
...
Mono.Cecil/Consts.cs(3,15): error CS0518: Predefined type 'System.String' is not defined or imported
# 7697 additional errors
```

This is kinda weird in that `mscorlib.dll` *is* referenced:

```
/reference:/Library/Frameworks/Mono.framework/Versions/Current/lib/mono/4.0-api/mscorlib.dll /reference:/Library/Frameworks/Mono.framework/Versions/Current/lib/mono/4.0-api/System.Core.dll ...
```

Equally odd is that a *subsequent* compile for `Mono.Cecil.dll` succeeded!

```
[2019-05-31T20:39:16.263Z] .../csc.exe ... /out:obj/net_4_0_Debug/net40/Mono.Cecil.dll
# no errors
```

Seems that the netstandard2.0 build is failing?

**Comment by jonpryor** [comment]
This might be a reason to "bite the bullet" and "binary rename" `Mono.Cecil.dll` to `Xamarin.Android.Cecil.dll`, instead of rebuilding from source.  However, I don't know if that's *actually* possible, nor how this would impact the Java.Interop build.

**Comment by jonpryor** [comment]
For comparison/reference, if I locally bump Java.Interop/external/cecil to https://github.com/mono/cecil/tree/a6c8f5e1070d25c38cbfaa415d6f6853e5236687, cecil builds without error when using mono 6.0.0.6.  I haven't tried using 6.4.0 yet.

**Comment by jonpryor** [comment]
For additional "giggles", this appears to be an msbuild problem.  As luck would have it, installing `MonoFramework-MDK-6.4.0.0.macos10.xamarin.universal.pkg` doesn't remove my previous mono install, which allows me to "replace" the new msbuild:

```bash
cd /Library/Frameworks/Mono.framework/Versions/6.4.0/lib/mono
sudo mv msbuild msbuild-6.4.0
sudo ln -s /Library/Frameworks/Mono.framework/Versions/6.0.0/lib/mono/msbuild .
```

So now I have msbuild 16.0.42-preview+g804bde742b (mono 6.0.0.6) instead of 16.0.457+g4e8391a0f3 (mono 6.4.0.0), while the rest of `Mono.framework` is 6.4.

Result?  `make all` within Java.Interop now completes without errors.

**Comment by jonpryor** [comment]
Since the current problems with this PR is that things can't even build with the updated `msbuild`, I've created a new PR which *just* bumps the submodule and doesn't bump the system mono version: PR #3159.

Let's see if that fares any better.

**Comment by lambdageek** [comment]
the `"GenerateDepsFile` msbuild issue is https://github.com/mono/mono/issues/13537

**Comment by jonpryor** [comment]
Apparently, the problem is that the dotnet SDK is "too old":

  * https://github.com/xamarin/monodevelop/pull/592
  * https://docs.microsoft.com/en-us/visualstudio/mac/net-core-support?view=vsmac-2019
  * https://xamarinhq.slack.com/archives/CDB6XAWLA/p1559594352078300

Part of the fix is to stop using such an old version of the dotnet SDK.  We need to ensure that dotnet 2.1.700 or later is installed, if *any* version is installed.

**Comment by jonpryor** [comment]
This PR should also add the "OOM-related test exclusions" from PR #3219:

https://github.com/xamarin/xamarin-android/pull/3219/files#diff-4fd6a6b4a0bb6335ccef7ff2cd35eedcR181
https://github.com/xamarin/xamarin-android/pull/3219/files#diff-c8f718ceb539a91456854e50089cb605R867

That way the xUnit tests will (hopefully) pass, which will give us a better feel for how things are doing.

**Comment by jonpryor** [comment]
Note that API Compatibility Checks are failing, e.g. https://jenkins.mono-project.com/job/xamarin-android-pr-pipeline-debug/486/API_20Compatibility_20Checks/

```html
<h3>Type Changed: System.Threading.Tasks.TaskExtensions</h3>
<p>Removed methods:</p>
<pre>
	<span class='removed removed-method breaking' data-is-breaking>public static System.Runtime.CompilerServices.ConfiguredCancelableAsyncEnumerable&lt;T&gt; ConfigureAwait&lt;T&gt; (this System.Collections.Generic.IAsyncEnumerable&lt;T&gt;, bool);</span>
	<span class='removed removed-method breaking' data-is-breaking>public static System.Runtime.CompilerServices.ConfiguredAsyncDisposable ConfigureAwait (this System.IAsyncDisposable, bool);</span>
	<span class='removed removed-method breaking' data-is-breaking>public static System.Runtime.CompilerServices.ConfiguredCancelableAsyncEnumerable&lt;T&gt; WithCancellation&lt;T&gt; (this System.Collections.Generic.IAsyncEnumerable&lt;T&gt;, System.Threading.CancellationToken);</span>
</pre>

<h2>Namespace System.Security.Cryptography</h2>
<h3>Removed Type <span class='breaking' data-is-breaking>System.Security.Cryptography.CryptographicOperations</span></h3>

<h2>Removed Namespace System.IO.Enumeration</h2>
<h3>Removed Type <span class='breaking' data-is-breaking>System.IO.Enumeration.FileSystemName</span></h3>
```

**Comment by jonpryor** [comment]
I don't think removing the timeout will actually work. :-(

https://github.com/xamarin/xamarin-android/pull/3155/commits/c8c21b3989f99c1e75a39aa19b2e6c548667b33d

The tests started over 4h ago, and they're still "running."  *However*:

I `ssh`ed into the machine running the tests & emulator.  They're running:

```
$ ~/android-toolchain/sdk/platform-tools/adb shell ps | grep Bcl
u0_a97       29336  1724 1423108 459088 0                   0 S Xamarin.Android.Bcl_Tests
```

The current time on the emulator:

```
$ ~/android-toolchain/sdk/platform-tools/adb shell date
Sat Jun 15 19:32:47 UTC 2019
```

Now, what's the emulator *doing*?

```
$ ~/android-toolchain/sdk/platform-tools/adb logcat -d > logcat2.txt
$ grep 29336 logcat2.txt
...
06-15 15:07:43.370 29336 29593 V xUnit   :    Test name: System.Net.Security.Tests.ServerRequireEncryptionTest.ServerRequireEncryption_ClientRequireEncryption_ConnectWithEncryption
```

The last message written from pid 29336  is from ~4h ago, at 15:07:43.370, vs "now", at 19:32:47.

The test process is still running, but apparently nothing is *happening*.  It's gone.

@grendello, @radekdoulik: "silly" idea?  Could we update `.apk` tests so that when a timeout occurs, we send a broadcast to the process which causes it to create a (new?) NUnit TestResults.xml file and print out the contents, *without* blocking for "all" tests to complete?  So that we can get *something*, instead of nothing?

**Comment by EgorBo** [comment]
Regarding the API breakage: https://github.com/dotnet/standard/commit/ebeec50871621ac116b1257c791fc162a0ac5ea0#diff-93ef3714cd10edf17d6d8045cf3406b8 (corresponding change in NS2.1)

**Comment by grendello** [comment]
> I don't think removing the timeout will actually work. :-(
> 
> [c8c21b3](https://github.com/xamarin/xamarin-android/commit/c8c21b3989f99c1e75a39aa19b2e6c548667b33d)
> 
> The tests started over 4h ago, and they're still "running." _However_:
> 
> I `ssh`ed into the machine running the tests & emulator. They're running:
> 
> ```
> $ ~/android-toolchain/sdk/platform-tools/adb shell ps | grep Bcl
> u0_a97       29336  1724 1423108 459088 0                   0 S Xamarin.Android.Bcl_Tests
> ```
> 
> The current time on the emulator:
> 
> ```
> $ ~/android-toolchain/sdk/platform-tools/adb shell date
> Sat Jun 15 19:32:47 UTC 2019
> ```
> 
> Now, what's the emulator _doing_?
> 
> ```
> $ ~/android-toolchain/sdk/platform-tools/adb logcat -d > logcat2.txt
> $ grep 29336 logcat2.txt
> ...
> 06-15 15:07:43.370 29336 29593 V xUnit   :    Test name: System.Net.Security.Tests.ServerRequireEncryptionTest.ServerRequireEncryption_ClientRequireEncryption_ConnectWithEncryption
> ```
> 
> The last message written from pid 29336 is from ~4h ago, at 15:07:43.370, vs "now", at 19:32:47.
> 
> The test process is still running, but apparently nothing is _happening_. It's gone.
> 
> @grendello, @radekdoulik: "silly" idea? Could we update `.apk` tests so that when a timeout occurs, we send a broadcast to the process which causes it to create a (new?) NUnit TestResults.xml file and print out the contents, _without_ blocking for "all" tests to complete? So that we can get _something_, instead of nothing?

I had a different thought once. What if, instead of what we do now, we wrote a partial result whenever any single test is finished? We could put the "fragments" into a directory and then merge them on the host into the XML. It might be easier to do it this way than taking the broadcast approach, I think

**Comment by radekdoulik** [comment]
I was thinking about that too. We might even write the xml file after each test with the results so far and set the `test-suite` result attribute `Failure` until we reach the last test. Don't know the details of xUnit/nUnit though.

Not sure what would it get us, as we will just know results of the tests that already run and nothing about the reasons of the timeout or crash. The adb logcat probably already has that information (until it is too big and shortened)

**Comment by steveisok** [comment]
I am closing this PR as I messed up the merge.  Please use https://github.com/xamarin/xamarin-android/pull/3246 for mono-2019-06 work

**Comment by steveisok** [comment]
@jonpryor On 6/15 mono got bumped to a branch that seemed to be from 2019-02. 

https://github.com/mono/mono/commit/52f6ea14

Was this expected?  Is that also why the MonoRequiredMin/MaxVersions went down to 6.0.0 and 6.1.0 in Configuration.props?  

Would this, in part, explain the remap-assembly-ref error we're seeing and which https://github.com/xamarin/xamarin-android/pull/3248 is trying to fix?

File: src/Mono.Android/Test/Xamarin.Android.Net/AndroidClientHandlerTests.cs [hunk 1]
```diff
@@ -79,7 +79,6 @@ public void Properties_Defaults ()
 			Assert.AreEqual (4096, h.CookieContainer.MaxCookieSize, "#3b");
 			Assert.AreEqual (null, h.Credentials, "#4");
 			Assert.AreEqual (50, h.MaxAutomaticRedirections, "#5");
-			Assert.AreEqual (int.MaxValue, h.MaxRequestContentBufferSize, "#6");
```
Comment (radekdoulik) [inline src/Mono.Android/Test/Xamarin.Android.Net/AndroidClientHandlerTests.cs]: @baulig Could you please shed some more light on why we return 0 for `MaxRequestContentBufferSize` property? I read the explanation in your [code](https://github.com/mono/mono/commit/2f411750acfd2132f9b4970ef8cf11d9a0f62df9#diff-a94223e905e684e2f789149da44fafd9R16), but didn't find any documentation about 0 meaning no limit. It only talks about the default value being 2 gigabytes, which is not the case now.

Should we really drop that check here on line 82?

**Comment by jonpryor** [comment]
Regarding the unit test failures in [`HttpClientHandlerTestBase.Properties_Defaults()`](https://jenkins.mono-project.com/job/xamarin-android-pr-pipeline-release/1504/testReport/(root)/AndroidClientHandlerTests/Mono_Android_Tests__Xamarin_Android_NetTests_AndroidClientHandlerTests_HttpClientHandlerTestBase_Properties_Defaults___Release_2/), which fails because `HttpClientHandler.MaxRequestContentBufferSize` returns 0 instead of `int.MaxValue`, that's because mono/2019-06 now uses the corefx `HttpClientHandler` codebase, which (apparently) has "deprecated" this property: https://github.com/dotnet/corefx/commit/e740973fed660da4426218bf2ec4be35a60ad095

> The HttpClientHandler property, MaxRequestContentBufferSize, has been
> deprecated.

Mono's solution to this change was to remove the assert:

https://github.com/mono/mono/commit/2f411750acfd2132f9b4970ef8cf11d9a0f62df9#diff-c2819b4b4d71d387feade1ab7e6ca795L76

We should either follow suit or change the test to assert that `HttpClientHandler.MaxRequestContentBufferSize` is now 0 instead of `int.MaxValue`.

See also: https://xamarinhq.slack.com/archives/C03CELHQQ/p1561742784381000

**Comment by baulig** [comment]
About `HttpClientIntegrationTests.Send_Transfer_Encoding_Custom()`: that test has been copied from Mono, but without the conditional that disables it when using the CoreFX handler.

However, the test is badly broken in itself and should be removed completely.

There is now a Mono PR up for this: https://github.com/mono/mono/pull/15470 and there will be two new xUnit tests that I'm going to upstream here: https://github.com/dotnet/corefx/pull/39039.

**Comment by jonpryor** [comment]
@baulig wrote:

> However, the test is badly broken in itself and should be removed completely.

How so?  [Mono PR #15470](https://github.com/mono/mono/pull/15470) implies that the problem is that it:

> didn't make much sense to use a `Transfer-Encoding` header with a `GET` request.

RFC 2616 does not stipulate that `Transfer-Encoding` is only for use with POST/non-`GET` requests, e.g. [RFC 2617 §3.6 Transfer Codings](https://www.w3.org/Protocols/rfc2616/rfc2616-sec3.html#sec3.6).

Perhaps the problem is the use of a `chunked2` header value?  This doesn't appear to actually be meaningful, in that the mono test is the only place that string appears:

```bash
# within mono/2019-02:
$ git grep chunked2
mcs/class/System.Net.Http/Test/System.Net.Http/HttpClientTest.cs:                               client.DefaultRequestHeaders.TransferEncoding.Add (new TransferCodingHeaderValue ("chunked2"));
```

Meanwhile, the [Mozilla docs for the `Transfer-Encoding` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding) also contain an example for [Chunked encoding](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding#Chunked_encoding), stating:

> Chunked encoding is useful when larger amounts of data are sent to the client and the total size of the response may not be known until the request has been fully processed. For example, when generating a large HTML table resulting from a database query or when transmitting large images.

This certainly implies that it *would* be useful in `GET` requests, from server to client.

*Furthermore*, https://github.com/dotnet/corefx/pull/39039 was closed without merging, so it appears that the end result is that the existing (previously passing!) tests have been *removed*, with no replacement in sight, with a reason that is unsatisfying.

**Comment by baulig** [comment]
> _Furthermore_, [dotnet/corefx#39039](https://github.com/dotnet/corefx/pull/39039) was closed without merging, so it appears that the end result is that the existing (previously passing!) tests have been _removed_, with no replacement in sight, with a reason that is unsatisfying.

I closed that PR because the CI pass revealed that the behavior is undefined - Windows, Linux and Mac all behave differently in CoreFX.

Whether or not an existing test passed or not is also mostly irrelevant when there are strong indications that the tested behavior might be incorrect.  In such situations, it's better to look at what the code is actually supposed to do, then check what it actually does, then revisit that test and carefully evaluate whether the tested behavior actually matches any of that.

If I understand this correctly, then the normal usage of the `HttpRequestHeaders.TransferEncoding` API is to tell `HttpClient` how to encode a `HttpContent` object that you pass to it.   It understands values such as for instance `"chunked"` and automatically enables chunked encoding if you do so.  But what should it do with a custom string such as `"chunked2"` - it knows nothing about such an encoding, so it could not actually "encode" that `HttpContent` with it.

This is also different from passing an arbitrary header (which is something that the CoreFX implementation is actually quite strict about) - this property actually enables a specific encoding, thus modifying how the `HttpContent` should be transmitted.

---

The old Mono behavior was that the following would throw _when attempting to send the request_:

```
     client.DefaultRequestHeaders.TransferEncoding.Add (new TransferCodingHeaderValue ("chunked"));
```

you'd have to do this instead this:

```
     client.DefaultRequestHeaders.TransferEncodingChunked = true;
     client.DefaultRequestHeaders.TransferEncoding.Add (new TransferCodingHeaderValue ("chunked"));
```

In CoreFX, both statements are equivalent.  You could use both, but you don't have to.

Mono also never allowed the following (and again, this would throw when _attempting to send the request_ and not when setting this property - which again is something that CoreFX changed: you shouldn't be able assign an invalid value to a property just to get an exception somewhere later down the line):

```
     client.DefaultRequestHeaders.TransferEncoding.Add (new TransferCodingHeaderValue ("chunked2"));
```

In CoreFX, the behavior is undefined.  It throws (when sending the request) on Windows, uses chunked encoding on Linux and no encoding on the Mac.

---

Furthermore, Mono's old implementation wasn't even throwing that `InvalidOperationException` that the test is checking for in `HttpClient`, but in `HttpWebRequest` (remove the catch block to see the full exception):

```
1) Send_Transfer_Encoding_Custom (MonoTests.System.Net.Http.HttpClientTest.Send_Transfer_Encoding_Custom)
  System.AggregateException : One or more errors occurred. (TransferEncoding requires the SendChunked property to be set to true.)
 ----> System.InvalidOperationException : TransferEncoding requires the SendChunked property to be set to true.
```

which is here: 
https://github.com/mono/mono/blob/d956cf687cb39bab18194a7eea608becfbc372d5/mcs/class/System/System.Net/HttpWebRequest.cs#L1176-L1187

**Comment by jonpryor** [comment]
Then we should entirely remove `HttpClientIntegrationTests.Send_Transfer_Encoding_Custom()`.

File: tests/BCL-Tests/Xamarin.Android.Bcl-Tests/LinkerDescription.xml [hunk 2]
```diff
@@ -0,0 +1,52 @@
+<?xml version="1.0" encoding="UTF-8" ?>
```
Comment (akoeplinger) [inline tests/BCL-Tests/Xamarin.Android.Bcl-Tests/LinkerDescription.xml]: this is no longer needed as the test was fixed in latest mono 2019-06.

**Comment by radekdoulik** [comment]
@grendello so the debug build error is indeed happening in `MonoAotOffsetsDumper.exe`. I noticed you fixed the tool for `NDK r17+` in the past.

Could you please take a look at it?

I have found a way to reproduce it locally. Run `make ANDROID_NDK_VERSION=r20 $(PWD)/android-cross-arm-release/armv7-linux-android.h` in `xa/external/mono/sdks/builds` and it should fail with:

```
Processing triple: armv7-none-linux-androideabi
Error parsing '/Users/rodo/git/xa2/external/mono/mono/metadata/metadata-cross-helpers.c'
/Users/rodo/android-toolchain/ndk/sysroot/usr/include/bits/ioctl.h(60,5): error: conflicting types for 'ioctl'
Error parsing '/Users/rodo/git/xa2/external/mono/mono/mini/mini-cross-helpers.c'
/Users/rodo/android-toolchain/ndk/sysroot/usr/include/bits/ioctl.h(60,5): error: conflicting types for 'ioctl'
make: *** [/Users/rodo/git/xa2/external/mono/sdks/builds/android-cross-arm-release/armv7-linux-android.h] Error 1
```

**Comment by radekdoulik** [comment]
The windows build test fails is expected, https://github.com/xamarin/xamarin-android/pull/3155/checks?check_run_id=163295931. It is failing on master too and should be fixed once https://github.com/xamarin/xamarin-android/pull/3305 is finished.

**Comment by radekdoulik** [comment]
@steveisok Not sure about the remaining test failure in Release, https://jenkins.mono-project.com/job/xamarin-android-pr-pipeline-release/1582/testReport/junit/Test%20collection%20for%20System.Net.Security.Tests/ServerNoEncryptionTest/System_Net_Security_Tests_ServerNoEncryptionTest_ServerNoEncryption_ClientNoEncryption_ConnectWithNoEncryption___Release/.

Does it ring a bell? Do you think it is Android specific?

I am curious whether it will fail in debug once the build is fixed (to see whether it might be linker issue).

**Comment by steveisok** [comment]
@radekdoulik See https://github.com/xamarin/xamarin-android/pull/3155#issuecomment-502672705 re: the API breaking changes.

**Comment by radekdoulik** [comment]
@steveisok for the API changes, the comment you mentioned explains only the extension methods location change. It doesn't explain the 2 removed types and the removed constructor. What is the story about those? Is it a regression?

**Comment by EgorBo** [comment]
@radekdoulik the removed ctor is also a dotnet/standard driven change. 
Not sure about the two missing types - I do see them in mono-2019-06, e.g. here https://github.com/mono/mono/blob/2019-06/mcs/class/Facades/netstandard/TypeForwarders.cs#L2537 +  https://github.com/mono/mono/blob/c5ed9bfc0d1b393c1dc6d97fc3caf8b5b09ab443/mcs/class/corlib/corlib.dll.sources#L1412

**Comment by radekdoulik** [comment]
OK, so the constructor is this issue: https://github.com/dotnet/standard/issues/1171 and was done intentionally in mono.

I will check whether the remaining types are really removed.

**Comment by radekdoulik** [comment]
@steveisok @EgorBo So the types were removed from the `System.dll` assembly and are only present in the `mscorlib.dll`. Shouldn't we add type forwarders to the `System.dll` to preserve the API compatability?

**Comment by akoeplinger** [comment]
Both CryptographicOperations and FileSystemName are new types in .NET Standard 2.1. They were part of .NET Core before and made it into mono from there, but weren't in any shipping version of NS.

I'm not sure whether we added typeforwarders for similar cases before.

**Comment by grendello** [comment]
> ANDROID_NDK_VERSION=r20 $(PWD)/android-cross-arm-release/armv7-linux-android.h

I managed to reproduce it and here's the [fix/workaround](https://gist.github.com/grendello/42e02089061163efeed7353a44676bb1). The issue comes from CppSharp's lack of knowledge about the `__asm__` compiler intrinsic used in the header file mentioned in the error message. Abbreviated version:

```c
int ioctl(int __fd, int __request, ...);

#if !defined(BIONIC_IOCTL_NO_SIGNEDNESS_OVERLOAD)
/* enable_if(1) just exists to break overloading ties. */
int ioctl(int __fd, unsigned __request, ...) __overloadable __enable_if(1, "") __RENAME(ioctl);
#endif
```

`RENAME` is defined as:

```c
#define __RENAME(x) __asm__(#x)
```

The real fix would have to be to make `CppSharp` understand the semantics of `__asm__` but it's way beyond the scope of this PR so the workaround works fine, as removing the `ioctl` definition out of the picture changes nothing from our point of view.

**Comment by radekdoulik** [comment]
@steveisok @EgorBo @akoeplinger I have checked both types and the https://raw.githubusercontent.com/xamarin/xamarin-android-api-compatibility/d16-1/reference/System.xml already contains `System.IO.Enumeration.FileSystemName`. So I think we should add type forwarder for it.

The other one, `System.Security.Cryptography.CryptographicOperations` is only present in https://raw.githubusercontent.com/xamarin/xamarin-android-api-compatibility/d16-2/reference/System.xml. It might be hard to remove it at this point, so I think we should rather add the forwarder as well.

**Comment by radekdoulik** [comment]
OTOH, the types are still present in `mscrolib`, so besides scenarios involving reflection, it should be relatively safe to remove them? Do you know any other cases, where it might break customer projects?

**Comment by steveisok** [comment]
The latest mono bump has the type forwards and should resolve the ConnectWithNoEncryption test

**Comment by radekdoulik** [comment]
The jenkins tests are green, azure test has less fails than master => ready to merge.

I am just going to reset the tests timeout values and merge.

**Comment by radekdoulik** [comment]
It turned out some of the submodules were on lower versions than master, so I reset the timeouts and updated the submodules.

Let see if the submodule bumps didn't break anything.

**Pull request merged**